### ISSUE TITLE
rustc hangs in rustc_trans::collector::collect_items_rec (or it's callees)

### ISSUE THREAD

**Issue by TimNN** [rust-lang/rust#37779]
As reported on reddit: https://www.reddit.com/r/rust/comments/5d1dzx/rustc_1130_and_nightly_infinite_loops_while/

Backtrace: https://gist.github.com/TimNN/6070f0da6e697a93ef13ab74d316524a

To reproduce, clone `clap-rs` from https://github.com/TimNN/clap-rs.git and check out the `hang` branch, which has the patch from the reddit thread applied.

**Comment by arielb1** [comment]
Looks like a duplicate of #37311.

**Comment by arielb1** [comment]
It's not, but rather an exponential case in trait selection for cases like
```
INFO:rustc_trans::common: Cache miss: Binder(<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<&mut std::env::ArgsOs>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:633:42: 633:54]> as std::iter::IntoIterator>) => VtableImpl(impl_def_id=DefId { krate: CrateNum(2), node: DefIndex(4170) => core/75fe2fdafd75f237ff013f4ea8c6fac39cd701ee2ed7455130a3f74d5902d10c::iter[0]::traits[0]::{{impl}}[0] }, substs=Slice([std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<std::iter::Map<&mut std::iter::Peekable<&mut std::env::ArgsOs>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:721:32: 721:64]>>, [closure@src/app/parser.rs:633:42: 633:54]>]), nested=[(), ()])
```

**Comment by michaelwoerister** [comment]
Thanks for copying the bug report over from reddit, @TimNN!

**Comment by ishitatsuyuki** [comment]
Should be fixed in #48296.
